[multistage] Initial (phase 1) Query runtime for  window functions - empty OVER() and OVER(PARTITION BY)

somandal opened the pull request:
This PR introduces the multi-stage runtime changes to support Phase 1 of Window Functions, specifically targeting empty `OVER()` and `OVER(PARTITION BY)`. Runtime support has been added for the following classes of window queries:

- Empty `OVER()` - single and multiple empty `OVER()`s in the same query
- `OVER(PARTITION BY)` - single and multiple `OVER(PARTITION BY)` using the same `PARTITION BY` key in the same query

This PR also fixes some bugs in the window function planning, specifically:

- Create Empty `OVER()` and `OVER(ORDER BY)` as singleton stages which need a single node assignment (similar to global aggregation without group by and global sort)
- ~Add some logic for `OVER(PARTITION BY key1 ORDER BY key1)` queries which get marked as `OVER(PARTITION BY key1)` to ensure that the `ORDER BY` key1's direction and null direction are the defaults, otherwise treat these as `OVER(PARTITION BY key1 ORDER BY key2)` type queries during planning.~

The window functions supported as part of Phase 1 are: SUM, AVG, MIN, MAX, COUNT, BOOL_OR, and BOOL_AND.
PostgreSQL supports many aggregation functions inside window functions as mentioned in their [documentation](https://www.postgresql.org/docs/current/functions-window.html):

> In addition to these functions, any built-in or user-defined ordinary aggregate (i.e., not ordered-set or hypothetical-set aggregates) can be used as a window function; see Section 9.21 for a list of the built-in aggregates. Aggregate functions act as window functions only when an OVER clause follows the call; otherwise they act as plain aggregates and return a single row for the entire set.

This PR also adds a new JSON file for window function query runtime tests via the output mechanism (H2 doesn't support window functions and got parse errors when trying to use it).

Design document: https://docs.google.com/document/d/13CmFm4djI09JKF_Xty5acoXxJoxC9CLXAsmikgzgtIs/edit?usp=sharing
OSS issue: https://github.com/apache/pinot/issues/7213
Window Function Planner PR: https://github.com/apache/pinot/pull/10228

This PR does not include support for:

- Execution engine changes for Phase 1 for `OVER(ORDER BY)` and `OVER (PARTITION BY ORDER BY)` (will be the next PR for this feature)
- Custom frames
- Other window functions related to rank and values
- Multiple window groups (basically multiple `OVER` clauses with different `PARTITION BY`, `ORDER BY` and/or `FRAME` specifications, if these specifications are the same they get grouped into a single window group)

The above will be part of future changes for window function support

cc @siddharthteotia @walterddr @vvivekiyer @Jackie-Jiang

somandal commented on pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotWindowExchangeNodeInsertRule.java:
```diff
@@ -160,6 +162,19 @@ private boolean isPartitionByOnlyQuery(Window.Group windowGroup) {
       Set<Integer> partitionByKeyList = new HashSet<>(windowGroup.keys.toList());
       Set<Integer> orderByKeyList = new HashSet<>(windowGroup.orderKeys.getKeys());
       isPartitionByOnly = partitionByKeyList.equals(orderByKeyList);
+      if (isPartitionByOnly) {
```
I'm not entirely sure if this is actually required, but thought I'd look into the correct semantics once I start the work on `ORDER BY` within `OVER()`. This check will prevent marking queries like below as partition by only queries:

- OVER(PARTITION BY key1 ORDER BY key1 DESC)
- OVER(PARTITION BY key1 ORDER BY key1 NULLS FIRST)

The code will however allow queries of the type `OVER(PARTITION BY key1 ORDER BY key1)` to be marked as partition by only (since all aggregated values will be the same across the whole partition just like OVER(PARTITION BY)).

Let me know if the preference is to just remove this extra check since logically the actual aggregation values will still be the same for the full partition irrespective of the direction and null direction.

walterddr replied:
all 3 of the example above produces the exact same query results right?

somandal replied:
yup, i tried these in postgres, but since postgres does an explicit sort for even the PARTITION BY keys it did make a difference in the ordering.

Since we've decided not to care about ordering on PARTITION BY key, the above 3 queries will have the same aggregation value. The NULLS LAST vs. NULLS FIRST will only modify the order of the NULL rows, but this can be handled within the operator and still treat this type of query as a PARTITION BY only query.

I guess based on my own response above, adding this check probably doesn't make sense 😅

somandal replied:
@walterddr should I go ahead and remove this check?

vvivekiyer replied:
+1 to removing this because:
1. We use isPartitionByOnlyQuery to decide if we have to use LogicalExchange vs LogicalSortExchange
2. Based on my understanding, no sorting of rows within a partition is needed if the partition key and order by key are the same.

somandal replied:
done, removed this check. I'll add special handling for null direction later on (left a TODO). This will only affect the order in which the columns with NULL values are output in the final result.

ankitsultana reviewed (commented):
Excited to see this feature. Left some comments based on a cursory look.

somandal commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/WindowAggregateOperator.java:
```diff
@@ -0,0 +1,431 @@
+      }
+    }
+    _hasReturnedWindowAggregateBlock = true;
+    if (rows.size() == 0) {
```
thanks for the suggestion @ankitsultana. Based on the plans I see that the aggregation and the aggregations with group by occur downstream (under) from the window operator. Is this still a concern in that case? Are there specific types of queries you think I should test out to check which behavior makes more sense here (`EOS` vs. empty `TransferableBlock`?

I've added a few tests in my latest commit to check scenarios like empty filter or empty filter with a GROUP BY (validated that these return an empty result). Are these scenarios relevant to your suggestion here?

ankitsultana replied:
Users can have a query where they first perform windowing in a sub-query, and then later perform aggregation with some filtering. Taking a Uber themed example, say there's a table which has an entry for each Eats order `(order_id, price,  category=[chinese, indian, italian, etc.], food_item=[burger, pizza, etc.])`, a query may be to find how many categories are there such that the second highest order food_item per category has more than 10k orders in the last day.

If there are no such food categories, ideally we should return a single row `[(0)]`, but if we don't return any block above the window operator we may end up returning no rows `[]`. (@walterddr can perhaps confirm this since he had fixed a related issue with empty response in agg operators)

Though this is a very specific example, I think it might be worthwhile to keep an invariant that each Operator would process at least one block.

somandal replied:
Got it, thanks for explaining the context here. I did look at @walterddr's change to make agg operators return a non-EOS response when rows is empty. This is only done when the group by set is empty (i.e. aggregation without group by query).

```
    if (rows.size() == 0) {
      if (_groupSet.size() == 0) {
        return constructEmptyAggResultBlock();
      } else {
        return TransferableBlockUtils.getEndOfStreamTransferableBlock();
      }
    } else {
      return new TransferableBlock(rows, _resultSchema, DataBlock.Type.ROW);
    }
```

My worry here when it come to returning an empty block is what should I fill in for the values of other columns (e.g. the select columns) for this scenario? Or do you think perhaps we should only return an empty block if there are no other input columns? cc @walterddr for any thoughts on this as well.

e.g. query where I need to worry about what to fill in for `a.col1`:

`SELECT a.col1, SUM(a.col3) OVER(PARTITION BY a.col1) FROM table a;`

ankitsultana replied:
I think you can pass in an empty container. `TransferableBlock(Collections.emptyList(), _resultSchema, DataBlock.Type.ROW)`.

somandal replied:
thanks for clarifying, done!

walterddr replied:
i am not sure i follow the discussion here. what exact query are we trying to mitigate?
using postgres, none of these query returns a single row
```
SELECT id, SUM(val) OVER() FROM test where 0 = 1
SELECT id, SUM(val) OVER(PARTITION BY id) FROM test where 0 = 1
SELECT id, SUM(val) OVER(PARTITION BY id ORDER BY val) FROM test where 0 =1
SELECT id, SUM(val) OVER(PARTITION BY id ORDER BY id) FROM test where 0 = 1
```

somandal replied:
I dug into this a bit more. @ankitsultana please confirm if the scenario I'm going to discuss here matches the scenario you mentioned or not.

So from what I understood, @ankitsultana wanted an empty block to be returned for queries with sub queries with the window function inside the sub query such as:

```
      {
        "sql": "SELECT SUM(count) FROM (SELECT string_col, COUNT(bool_col) OVER() as count FROM {tbl} WHERE string_col = 'a' AND bool_col = false AND int_col > 200)",
        "outputs": [
          [0]
        ]
      },
```

In scenarios where the window inside the sub-query returns 0 rows, the outer query's aggregation should still work and return a result of '0' instead of no rows.

Now I actually tried the above (and will add tests with my latest commit) and found that it doesn't matter if I return an empty row or EOS here. In both scenarios the outer query aggregation returns the correct result.

Based on this @ankitsultana and @walterddr can you confirm if I can change this back to return EOS when the number of rows is 0?

ankitsultana replied:
I see. In that case either ways works for me. Can we also add this to the documentation of `MultistageOperator::nextBlock`?

somandal replied:
I'm not sure if adding documentation on this is really necessary. The above works because the outer query uses an intermediate AggregateOperator which does the right thing (i.e. @walterddr 's fix that we discussed above).

walterddr replied:
correct. the tradition of Postgres (as far as I know), is that all aggregate without group should return 1 row. for example
```
SELECT * FROM tbl WHERE val > (SELECT MAX(val) FROM tbl2)
SELECT * FROM tbl WHERE (SELECT count(val) FROM tbl2) > 0
```
both example the SQL relational planner interprets the inner query result as a single value. this can only happen if the agg without group by operators as I mentioned above.

So i think this rule doesn't apply to window, please share more if the analysis above can be in any way improved

somandal replied:
Yup makes sense! I've changed this back to return an `EOS` since this doesn't apply to window functions and the `AggregateOperator` handles this case.

somandal commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/WindowAggregateOperator.java:
```diff
@@ -0,0 +1,431 @@
+  private boolean isPartitionByOnlyQuery(List<RexExpression> groupSet, List<RexExpression> orderSet,
+      List<RelFieldCollation.Direction> orderSetDirection,
+      List<RelFieldCollation.NullDirection> orderSetNullDirection) {
+    if (orderSet == null || orderSet.isEmpty()) {
```
done

somandal commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/WindowAggregateOperator.java:
```diff
@@ -0,0 +1,431 @@
+        // TODO: Move to close call.
+        return TransferableBlockUtils.getEndOfStreamTransferableBlock();
+      }
+    } catch (Exception e) {
```
done

ankitsultana commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/WindowAggregateOperator.java:
```diff
@@ -0,0 +1,436 @@
+  @Override
+  protected TransferableBlock getNextBlock() {
+    try {
+      if (!_readyToConstruct && !consumeInputBlocks()) {
```
nit: I think `_readyToConstruct` is not really needed since `consumeInputBlocks` indicates that no more blocks need to be consumed, and in case of error the `_upstreamErrorBlock` would be set.

somandal replied:
Agreed that `_readyToConstruct` is not needed as such for correctness. Without using `_readyToConstruct` an extra call to `consumeInputBlocks` may be required to return the `EOS` (since today `_readyToConstruct` caches that EOS has been seen already). Not that this is a big deal, but might be okay to keep this as is for now.

walterddr replied:
let's keep it this way. 
the `readyToConstruct` flag is only here to ensure correctness. for example if the data is already sorted and we are doing partition by order by, then we can actually produce rows as we go. 

let's revisit this later

ankitsultana reviewed (approved):
left a nit but lgtm. Thanks for addressing the feedback.

walterddr commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java:
```diff
@@ -308,6 +303,20 @@ public Object merge(Object agg, Object value) {
+
+    @Override
+    public Object merge(Object left, Object ignored) {
+      // TODO: COUNT(*) doesn't need to parse right object until we support NULL
+      return ((Number) left).doubleValue() + 1;
```
this should be supported right?
```suggestion
      return ((Number) left).doubleValue() + (ignored == null ? 0 : 1);
```
and change `ignored` to `right`

somandal replied:
done

walterddr commented on pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTestBase.java:
```diff
@@ -364,6 +371,8 @@ public static class Query {
       public List<List<Object>> _outputs = null;
       @JsonProperty("expectedException")
       public String _expectedException;
+      @JsonProperty("keepOutputRowOrder")
+      public boolean _keepOutputRowOrder;
```
good catch. would you be able to also change the OrderBy.json and add the keepOutputRowOrder flag?

somandal replied:
done, ran into an issue with a few tests and left TODOs to look into them later along with comments explaining what I think is going on. As a follow up perhaps I can revisit making this 'keepOutputRowOrder' less flaky in terms of ordering (i.e. handle scenarios where the order by key is correctly ordered but there are ties)

walterddr commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/WindowAggregateOperator.java:
```diff
@@ -0,0 +1,436 @@
+    Object merge(Object agg, Object value);
+  }
+
+  private static class WindowAccumulator {
+    private static final Map<String, Function<DataSchema.ColumnDataType, WindowMerger>> WINDOW_MERGERS =
+        ImmutableMap.<String, Function<DataSchema.ColumnDataType, WindowMerger>>builder()
+            .put("SUM", cdt -> WindowAggregateOperator::mergeSum)
+            .put("$SUM", cdt -> WindowAggregateOperator::mergeSum)
+            .put("$SUM0", cdt -> WindowAggregateOperator::mergeSum)
+            .put("MIN", cdt -> WindowAggregateOperator::mergeMin)
+            .put("$MIN", cdt -> WindowAggregateOperator::mergeMin)
+            .put("$MIN0", cdt -> WindowAggregateOperator::mergeMin)
+            .put("MAX", cdt -> WindowAggregateOperator::mergeMax)
+            .put("$MAX", cdt -> WindowAggregateOperator::mergeMax)
+            .put("$MAX0", cdt -> WindowAggregateOperator::mergeMax)
+            .put("COUNT", cdt -> new MergeCount())
+            .build();
```
let's extract out these to a common `Accumulator` interface and reused in both agg operator and window operator

somandal replied:
done. should have done this right from the get go :)

walterddr commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/WindowAggregateOperator.java:
```diff
@@ -0,0 +1,436 @@
+    }
+  }
+
+  private static class WindowFrame {
```
javadoc please, i am not sure what isRows mean --> are we using this to determine this RANGE-based or ROW-based?

somandal replied:
done. yes your understanding of isRows is indeed ROWS vs. RANGE

walterddr commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/WindowAggregateOperator.java:
```diff
@@ -0,0 +1,436 @@
+    return false;
+  }
+
+  private static Key extractRowKey(Object[] row, List<RexExpression> groupSet) {
```
same here please make a comment Agg util for keeping these utils 
we can even have a separate PR to just refactor these out

somandal replied:
done

vvivekiyer commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/WindowAggregateOperator.java:
```diff
@@ -0,0 +1,343 @@
+      return true;
+    }
+
+    if (CollectionUtils.isEmpty(groupSet) || (groupSet.size() != orderSet.size())) {
```
In Phase1 (not specific to this PR), are we going to support queries where there are multiple window functions but one function has an ORDER BY and the other doesn't? I'm assuming it fails the criteria - "All order by keys should be on a single column" and Calcite will generate 2 window groups. Is that correct?

`SELECT SUM(col1) OVER(PARTITION BY key1), SUM(col2) OVER(PARTITION BY key1 ORDER BY col1) from table t1`

Either way, If we don't have a test for this case, can we please add them

somandal replied:
Yes you're right, such queries will result in more than 1 window group and is not supported as part of Phase 1. I've enhanced the planner tests to add more scenarios for multiple window groups (had just 1 earlier).

walterddr reviewed (commented):
lgtm overall. tagging @xiangfu0 and @siddharthteotia for another look

somandal commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/WindowAggregateOperator.java:
```diff
@@ -0,0 +1,343 @@
+ * columns and in addition will add the aggregation columns to the output data.
+ * [input columns, aggregate result1, ... aggregate resultN]
+ *
+ * The window functions supported today are SUM/COUNT/MIN/MAX aggregations. Window functions also include
```
Yes correct, added AVG to the list

somandal commented on pinot-query-planner/src/main/java/org/apache/pinot/query/planner/stage/WindowNode.java:
```diff
@@ -20,6 +20,7 @@
 
 import com.clearspring.analytics.util.Preconditions;
 import java.util.ArrayList;
+import java.util.Collections;
```
Agreed it's not needed here. Removed it

vvivekiyer reviewed (approved):
LGTM.

siddharthteotia commented:
Will be reviewing this today

siddharthteotia commented on pinot-query-planner/src/main/java/org/apache/pinot/query/planner/StageMetadata.java:
```diff
@@ -81,6 +82,12 @@ public void attach(StageNode stageNode) {
     }
+    if (stageNode instanceof WindowNode) {
+      WindowNode windowNode = (WindowNode) stageNode;
+      // Empty OVER() and OVER(ORDER BY) need to be processed on a singleton node. OVER() with PARTITION BY can be
```
> Empty OVER() and OVER(ORDER BY) need to be processed on a singleton node.

Add a TODO to improve this in future per original discussion ?

somandal replied:
done

siddharthteotia commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/WindowAggregateOperator.java:
```diff
@@ -0,0 +1,332 @@
+ * will output as many rows as input rows.
+ *
+ * Note: This class performs aggregation over the double value of input.
+ * If the input is single value, the output type will be input type. Otherwise, the output type will be double.
```
Not sure I follow this. What is the special casing here w.r.t single value v/s multi value ?

somandal replied:
This comment is copied over from `AggregateOperator` (since both use the `AggregationUtils`) and basically discusses that if only a single input is part of the aggregation, then the `Merger::initialize()` function today doesn't cast it as a double, so the output won't be a double. For other cases it does use the double value:

```
    default Object initialize(Object other, DataSchema.ColumnDataType dataType) {
      // TODO: Initialize as a double so that if only one row is returned it matches the type when many rows are
      //       returned
      return other == null ? dataType.getNullPlaceholder() : other;
    }
```

Example merger function:

```
  private static Object mergeMax(Object left, Object right) {
    return Math.max(((Number) left).doubleValue(), ((Number) right).doubleValue());
  }
```

I tried to fix this but ran into a bunch of test issues so decided to try and fix the `initialize` to also use `doubleValue()` in a separate PR if possible. Also wanted context on why the code was written in this way. Hope this clarifies the comment.

somandal replied:
@walterddr quick question in case you have context. Why do we initialize the aggregation output without using `doubleValue()` but use `doubleValue()` in the aggregation? Just wanted to understand the context and see if I can make a fix for this in a follow up PR

siddharthteotia commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/WindowAggregateOperator.java:
```diff
@@ -0,0 +1,332 @@
+ * If the input is single value, the output type will be input type. Otherwise, the output type will be double.
+ *
+ * TODO:
+ *     1. Add support for OVER() clause with ORDER BY only or PARTITION BY ORDER BY
```
Do we need to add a TODO bullet point for multiple `OVER()` with different `PARTITION BY `and `ORDER BY` ?

On a related note, you may want to update the PR description that changes here also support multiple `OVER()` with `PARTITION BY` on the same column ?

somandal replied:
added a TODO for multiple window groups.
Technically the multiple window groups change should happen on the planning side where we create multiple `LogicalWindow` nodes in the tree. We can still keep a TODO here in case any operator changes are indeed needed.

siddharthteotia commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/WindowAggregateOperator.java:
```diff
@@ -0,0 +1,332 @@
+      List<RexExpression> orderSet, List<RelFieldCollation.Direction> orderSetDirection,
+      List<RelFieldCollation.NullDirection> orderSetNullDirection, List<RexExpression> aggCalls, int lowerBound,
+      int upperBound, boolean isRows, List<RexExpression> constants, DataSchema resultSchema, DataSchema inputSchema,
+      long requestId, int stageId, VirtualServerAddress virtualServerAddress) {
```
(nit) is it possible to encapsulate the input into an POJO type object ? Might look cleaner / easy to read / review. 

Not necessary for this PR though

somandal replied:
yeah sure let me explore this as a future improvement

siddharthteotia commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/WindowAggregateOperator.java:
```diff
@@ -0,0 +1,332 @@
+    _orderSetInfo = new OrderSetInfo(orderSet, orderSetDirection, orderSetNullDirection);
+    _windowFrame = new WindowFrame(lowerBound, upperBound, isRows);
+
+    // TODO: add support for custom frames, and for ORDER BY default frame (upperBound => currentRow)
```
(nit) might want to capture TODOs in one place.. seem to be repeated here as everything is already called out in the constructor JavaDoc ?

somandal replied:
done

siddharthteotia commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/WindowAggregateOperator.java:
```diff
@@ -0,0 +1,332 @@
+        "Only default frame is supported, upperBound must be UNBOUNDED FOLLOWING or CURRENT ROW");
+
+    // we expect all agg calls to be aggregate function calls
+    _aggCalls = aggCalls.stream().map(RexExpression.FunctionCall.class::cast).collect(Collectors.toList());
```
Will we throw compilation error if after merging this PR, someone attempts `RANK` or say `ROW_NUMBER`  or say value functions ?

siddharthteotia replied:
Also such error can be thrown sooner than operator creation point imo ?

somandal replied:
These checks are done in the planner while building the calcite plan. I have a planner test too to ensure that `ROW_NUMBER` etc throw.

siddharthteotia replied:
Great. Thanks

siddharthteotia commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/WindowAggregateOperator.java:
```diff
@@ -0,0 +1,332 @@
+    _windowFrame = new WindowFrame(lowerBound, upperBound, isRows);
+
+    // TODO: add support for custom frames, and for ORDER BY default frame (upperBound => currentRow)
+    Preconditions.checkState(!_windowFrame.isRows(), "Only RANGE type frames are supported at present");
```
Hmm not sure I understand this. I thought FRAME is not supported in the initial implementation based on the phases laid out in design doc.

somandal replied:
These checks are for default frames (just to ensure we aren't getting inputs where the frames are overridden)

- Without `ORDER BY`: default frame: `UNBOUNDED PRECEDING` to `UNBOUNDED FOLLOWING`
- With `ORDER BY`: default frame: `UNBOUNDED PRECEDING` to `CURRENT ROW`

Also whether `ROWS` or `RANGE` is the default depends on the default frame so that's why I check for that too.

siddharthteotia commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/WindowAggregateOperator.java:
```diff
@@ -0,0 +1,332 @@
+    Preconditions.checkState(_windowFrame.isUnboundedPreceding(),
+        "Only default frame is supported, lowerBound must be UNBOUNDED PRECEDING");
+    Preconditions.checkState(_windowFrame.isUnboundedFollowing()
+            || (_windowFrame.isUpperBoundCurrentRow() && isPartitionByOnly),
```
Why is 2nd check (after OR) needed ?

somandal replied:
This is a temporary check because today we don't support window functions with `ORDER BY` where the `ORDER BY` key is different from the `PARTITION BY` key. We do support queries where the keys are the same, e.g. `OVER(PARTITION BY key1 ORDER BY key1)`.

For queries such as `OVER(PARTITION BY key1 ORDER BY key1)`, the default frame lands up getting set to `UNBOUNDED PRECEDING` to `CURRENT ROW`. This check here ensures that if the `upperBound` is indeed `CURRENT ROW` we should check that it's  a query where the `PARTITION BY` and `ORDER BY` key is the same.

siddharthteotia commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/WindowAggregateOperator.java:
```diff
@@ -0,0 +1,332 @@
+    // TODO: Not all window functions (e.g. ROW_NUMBER, LAG, etc) need aggregations. Such functions should be handled
+    //       differently.
+    _windowAccumulators = new AggregationUtils.Accumulator[_aggCalls.size()];
+    for (int i = 0; i < _aggCalls.size(); i++) {
```
(nit) may be do `aggCalls.size()` once before the loop ?

somandal replied:
done

siddharthteotia commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/WindowAggregateOperator.java:
```diff
@@ -0,0 +1,332 @@
+      RexExpression.FunctionCall agg = _aggCalls.get(i);
+      String functionName = agg.getFunctionName();
+      if (!mergers.containsKey(functionName)) {
+        throw new IllegalStateException("Unexpected value: " + functionName);
```
So this is when anything other than SUM / MIN / MAX / COUNT / AVG is used ?

Might want to use a slightly more self understandable error message

somandal replied:
We also support BOOL_OR and BOOL_AND
Updated the error message

siddharthteotia commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/WindowAggregateOperator.java:
```diff
@@ -0,0 +1,332 @@
+      return _upperBound;
+    }
+  }
+}
```
Don't we also need to implement close() to cleanup any resources ? IIRC, recently all operators were changed to extend from `AutoCloseable` and therefore should try to override close() if needed

somandal replied:
Great question. All intermediate stage operators extend from `MultiStageOperator` which has the implementation for `close()` and `cancel()`. Most intermediate stage operators don't need to override these as the same behavior is required.

Code from `MultiStageOperator`:

```
  // TODO: Ideally close() call should finish within request deadline.
  // TODO: Consider passing deadline as part of the API.
  @Override
  public void close() {
    for (MultiStageOperator op : getChildOperators()) {
      try {
        op.close();
      } catch (Exception e) {
        LOGGER.error("Failed to close operator: " + op + " with exception:" + e);
        // Continue processing because even one operator failed to be close, we should still close the rest.
      }
    }
  }

  public void cancel(Throwable e) {
    for (MultiStageOperator op : getChildOperators()) {
      try {
        op.cancel(e);
      } catch (Exception e2) {
        LOGGER.error("Failed to cancel operator:" + op + "with error:" + e + " with exception:" + e2);
        // Continue processing because even one operator failed to be cancelled, we should still cancel the rest.
      }
    }
  }
```

siddharthteotia replied:
Got it. Makes sense

siddharthteotia commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/WindowAggregateOperator.java:
```diff
@@ -0,0 +1,332 @@
+    // The lower bound of the frame. Set to Integer.MIN_VALUE if UNBOUNDED PRECEDING
+    final int _lowerBound;
+    // The lower bound of the frame. Set to Integer.MAX_VALUE if UNBOUNDED FOLLOWING. Set to 0 if CURRENT ROW
+    final int _upperBound;
```
(nit) multi stage engine's query planning is likely to be heap heavy so we may want to double check if `int` is really needed or can we do away with `short`. 

not a big deal in this PR but something to do in follow-up if folks agree

somandal replied:
Great suggestion! I'll revisit these once I start work on frame support as I want to look into whether there is a recommendation on the frame limits or not. Will take this as an action item for frame support.

somandal replied:
Agree! we'll need to test this out for performance for sure

siddharthteotia commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/WindowAggregateOperator.java:
```diff
@@ -0,0 +1,332 @@
+        Key key = AggregationUtils.extractRowKey(row, _groupSet);
+        _partitionRows.putIfAbsent(key, new ArrayList<>());
+        _partitionRows.get(key).add(row);
+        for (int i = 0; i < _aggCalls.size(); i++) {
```
Suggest moving size() before loop

somandal replied:
done

siddharthteotia commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/WindowAggregateOperator.java:
```diff
@@ -0,0 +1,332 @@
+    return partitionByInputRefIndexes.equals(orderByInputRefIndexes);
+  }
+
+  private TransferableBlock produceWindowAggregateBlock() {
```
(nit) suggest renaming this to `produceOutputBlock`

somandal replied:
Keeping the name the same to be consistent with `AggregateOperator` and `SortOperator`

siddharthteotia commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/WindowAggregateOperator.java:
```diff
@@ -0,0 +1,332 @@
+        rows.add(row);
+      }
+    }
+    _hasReturnedWindowAggregateBlock = true;
```
The state that whether or not we can produce output block should be maintained and dictated by the piece of code that consumed the input block - `consumeInputBlocks`

Generally speaking, whenever an operator consumes an input block, before exiting that function it should be able to determine whether or not it is ready to produce output and accordingly change the local state which can then be leveraged in `getNextBlock`.

siddharthteotia replied:
^^ might just be more intuitive but nothing wrong with the current approach

somandal replied:
I've tried to keep the operator flow similar to `AggregateOperator` for now. This flag `_hasReturnedWindowAggregateBlock` is meant to differentiate the scenario where we finished consumption and haven't yet sent the output block, vs. the scenario where we finished consumption and sent the output block already but now just need to send a final `EOS`. In both cases the `consumeInputBlocks` will indicate there is nothing more to send.

We can revisit this logic for all operators at once if needed.

siddharthteotia replied:
Sounds good. FYI @walterddr

walterddr replied:
for now the logic is kind of there to guard the EOS as @somandal mentioned. we can revisit in general later. please kindly file an issue. (I am yet to understand what's the confusion and how can we improve it, but will look into it more generally once we file the issue with more context)

siddharthteotia commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/WindowAggregateOperator.java:
```diff
@@ -0,0 +1,332 @@
+      List<Object[]> rowList = e.getValue();
+      for (Object[] existingRow : rowList) {
+        Object[] row = new Object[existingRow.length + _aggCalls.size()];
+        System.arraycopy(existingRow, 0, row, 0, existingRow.length);
```
Not really related to this PR but probably something we can improve generally

I am thinking that if we do the processing of operators in off heap buffers, then can we do something like this to avoid memcpy ?

Say we setup 3 ops in an op chain

Op1 -> Op2 -> Op3 from upstream to downstream

When the operators are setup, they get an input container (aka record batch with schema)

Op3's input container = Op2's output container
Op2's input container = Op1's output container

When the `OpChain` is setup, we call something like setup() on operator

Output setup (Input) {
   // create Output container and return it
   // the caller will use the returned Output as Input o setup the next downstream operator
}

We produce into Output but that does not require memcpy since it is direct buffers

somandal replied:
Great suggestion. Let's look into this as a separate change.

siddharthteotia replied:
Sounds good. FYI @walterddr

walterddr replied:
i dont understand why we need to copy in the first place. window over is 1-in-1-out in the row perspective. it only caches the row it gets, accumulate the results for the additional columns, and attach the result back to the row when it finishes. 

i agree we will have overhead for heap memory b/c we need to buffer everything. but not sure why we need a mem copy, it felt like if we are using `List<Object>` instead of `Object[]` we can completely avoid this copy yes?

somandal replied:
We need to create an output in the format of `List<Object[]>`, where each `Object[]` contains the contents of a single row. The output `Object[]` row size is larger than the input row's size since we need to store the aggregated columns. Using `List<Object>` instead of `Object[]` can avoid the copy as it can be appended to and added to the output result.

somandal commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/WindowAggregateOperator.java:
```diff
@@ -0,0 +1,332 @@
+        //       handling for all query types
+        Key key = AggregationUtils.extractRowKey(row, _groupSet);
+        _partitionRows.putIfAbsent(key, new ArrayList<>());
+        _partitionRows.get(key).add(row);
```
done

siddharthteotia reviewed (approved):
LGTM. Yet to finish going through some of the test changes. Reviewed everything else.

walterddr reviewed (approved):
lgtm with some more comments. thank you @somandal for the contribution. this is a huge effort and definitely appreciate the patience and consideration during the review/comment process.

somandal commented on pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotWindowExchangeNodeInsertRule.java:
```diff
@@ -48,7 +48,7 @@ public class PinotWindowExchangeNodeInsertRule extends RelOptRule {
   // Supported window functions
   private static final Set<SqlKind> SUPPORTED_WINDOW_FUNCTION_KIND = ImmutableSet.of(SqlKind.SUM, SqlKind.SUM0,
-      SqlKind.MIN, SqlKind.MAX, SqlKind.COUNT);
+      SqlKind.MIN, SqlKind.MAX, SqlKind.COUNT, SqlKind.OTHER_FUNCTION);
```
Ah good point, I think we just saw that PostgreSQL has support for all aggregation functions and thought why not add what we have (at least the basic ones). Hope it's alright to leave this here. I've updated to add a comment here.

somandal commented on pinot-query-planner/src/main/java/org/apache/pinot/query/planner/stage/WindowNode.java:
```diff
@@ -79,7 +79,7 @@ public WindowNode(int stageId, List<Window.Group> windowGroups, List<RexLiteral>
     //       Frame literals come in the constants from the LogicalWindow and the bound.getOffset() stores the
     //       InputRef to the constants array offset by the input array length. These need to be extracted here and
     //       set to the bounds.
-    validateFrameBounds(windowGroup.lowerBound, windowGroup.upperBound, windowGroup.isRows);
```
done

somandal replied:
never mind, updated this to add the enum directly into `WindowNode` and access it in `WindowAggregateOperator`.

somandal commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/utils/AggregationUtils.java:
```diff
@@ -0,0 +1,179 @@
+ * Utility class to perform aggregations in the intermediate stage operators such as {@code AggregateOperator} and
+ * {@code WindowAggregateOperator}
+ */
+public class AggregationUtils {
```
done

somandal commented:
> lgtm with some more comments. thank you @somandal for the contribution. this is a huge effort and we definitely appreciate the patience and consideration during the review/comment process.

Thanks @walterddr! Appreciate the thorough review and the guidance on this from all involved.
I've addressed all comments. PTAL when you get a chance
